### PR TITLE
Provide MSB5009 error with project's name and it's GUID

### PR DESCRIPTION
**Pull request by BartoszKlonowski**
This pull request fixes: #4836

Changes introduced in this delivery extends the *MSB5009* with additional information about the incorrectly nested project within the solution. Additional details are:
* incorrectly nested project's name (without the full path, unique name only),
* incorrectly nested project's GUID

**NOTE:** original *MSB5009* error is kept in the resources for backward compatibility (see commit messages) while the extended MSB5009 can be easily invoked with extended name.

---

Changes done in this delivery are:
* Create additional version of *MSB5009* error with modified error message - meaning stays the same as original *MSB5009* error, but incorrectly nested project's name and GUID is added as arguments.
* Replace the old *MSB5009* with the extended one for each place where this extended error should be used.
NOTE: `SolutionFile.ParseNestedProjects()` method has the original error kept, as it does specify at least the line of the solution file with the incorrect project. Moreover, at the moment of throwing the error, there's no possibility of getting the name of the project.
* Cover the incorrect nesting case with the unit test checking whether for an original reproduction solution (please check the commit message for link) proper error with both name and GUID is thrown.

For more details about each change please see the commit message.

---

Example
of original *MSB5009* error vs the extended version presented below. For both cases the same solution file was used, which is the solution file mentioned before available to get from the original [issue](https://github.com/dotnet/msbuild/issues/4835).

The original error:
obraz
The extended version (the results of this following PR):
obraz

### CODE REVIEW

File: src/Shared/Resources/Strings.shared.resx [hunk 1]
```diff
@@ -200,6 +200,10 @@
     <value>MSB5008: Error parsing the solution configuration section in solution file. The entry "{0}" is invalid.</value>
     <comment>{StrBegin="MSB5008: "}UE: The solution filename is provided separately to loggers.</comment>
   </data>
+  <data name="SolutionParseNestedProjectErrorWithNameAndGuid" Visibility="Public">
+    <value>MSB5009: Error parsing the nested project "{0}" section with GUID: "{1}" in solution file.</value>
```
Comment (rainersigwald) [inline src/Shared/Resources/Strings.shared.resx:204]: Should this also pass (and report) `ParentProjectGuid`? Maybe something like

```suggestion
    <value>MSB5009: Error parsing the project "{0}" section with GUID: "{1}". It is nested under "{2}" but that project is not found in the solution.</value>
```

?

**Review by rainersigwald — COMMENTED** [review]
Thanks! Your test did what the other tests around here did, but Shouldly offers us a nicer way to do it, so I pushed a switch to that mechanism.

Including @KathleenDollard for opinions on the error message (in addition to mine).

File: src/Shared/Resources/Strings.shared.resx (hunk 1 above)
Comment (BenVillalobos) [reply src/Shared/Resources/Strings.shared.resx:204]: +1. Might as well display as much helpful information as possible.

File: src/Build.UnitTests/Construction/SolutionFile_Tests.cs [hunk 2]
```diff
@@ -1223,6 +1223,49 @@ public void MissingNestedProject()
...
+                        SolutionGuid = {AF600A67-B616-453E-9B27-4407D654F66E}
+                    EndGlobalSection
+                EndGlobal
+                ";
+
+            InvalidProjectFileException e = Should.Throw<InvalidProjectFileException>(() => ParseSolutionHelper(solutionFileContents));
+
+            e.ErrorCode.ShouldBe("MSB5009");
+            e.Message.ShouldContain( "{1484A47E-F4C5-4700-B13F-A2BDB6ADD35E}", e.Message);
+            e.Message.ShouldContain("ConsoleApp1", e.Message);
```
Comment (Forgind) [inline src/Build.UnitTests/Construction/SolutionFile_Tests.cs:1267]:
```suggestion
            e.Message.ShouldContain("{1484A47E-F4C5-4700-B13F-A2BDB6ADD35E}");
            e.Message.ShouldContain("ConsoleApp1");
```

**Review by Forgind — COMMENTED** [review]
Looks good to me! Pending string review.

File: src/Build/Construction/Solution/ProjectInSolution.cs [hunk 3]
```diff
@@ -370,7 +370,7 @@ internal string GetUniqueProjectName()
                         if (!ParentSolution.ProjectsByGuid.TryGetValue(ParentProjectGuid, out ProjectInSolution proj))
                         {
                             ProjectFileErrorUtilities.VerifyThrowInvalidProjectFile(proj != null, "SubCategoryForSolutionParsingErrors",
-                                new BuildEventFileInfo(ParentSolution.FullPath), "SolutionParseNestedProjectError");
+                                new BuildEventFileInfo(ParentSolution.FullPath), "SolutionParseNestedProjectErrorWithNameAndGuid", ProjectName, ProjectGuid);
```
Comment (Forgind) [inline src/Build/Construction/Solution/ProjectInSolution.cs:373]:
```suggestion
                                new BuildEventFileInfo(ParentSolution.FullPath), "SolutionParseNestedProjectErrorWithNameAndGuid", ProjectName, ProjectGuid, ParentProjectGuid);
```

File: src/Build/Construction/Solution/ProjectInSolution.cs [hunk 4]
```diff
@@ -409,7 +409,7 @@ internal string GetOriginalProjectName()
                         if (!ParentSolution.ProjectsByGuid.TryGetValue(ParentProjectGuid, out ProjectInSolution parent))
                         {
                             ProjectFileErrorUtilities.VerifyThrowInvalidProjectFile(parent != null, "SubCategoryForSolutionParsingErrors",
-                                new BuildEventFileInfo(ParentSolution.FullPath), "SolutionParseNestedProjectError");
+                                new BuildEventFileInfo(ParentSolution.FullPath), "SolutionParseNestedProjectErrorWithNameAndGuid", ProjectName, ProjectGuid);
```
Comment (Forgind) [inline src/Build/Construction/Solution/ProjectInSolution.cs:412]:
```suggestion
                                new BuildEventFileInfo(ParentSolution.FullPath), "SolutionParseNestedProjectErrorWithNameAndGuid", ProjectName, ProjectGuid, ParentProjectGuid);
```

**Review by Forgind — APPROVED** [review]
CI errors were just because the translations were out-of-date. You can fix that by building.

**Comment by BartoszKlonowski** [comment]
I was just about to deliver the review remarks, but thank you for the commitment!

**Comment by rainersigwald** [comment]
Thanks, @BartoszKlonowski

**Pull request merged**